Int b 22478 error requested pickup date past on edit

ryan-mchugh opened the pull request:
## [B-22478 - Disable non-PPM shipment submission when pickup date invalid](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-22478)

## Summary

We are adding validation for the `requestedPickupDate` when editing/creating shipments for ALL office users, and customer. `requestedPickupDate` must now be in the future when creating a shipment, and if modified when editing a shipment.

NOTE - validation for 22478 is UI only for a proactive error. API protections coming in downstream B-22978

## Verification Steps for the Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [ ] Have the Agility acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [ ] Has the branch been pulled in and checked out?
- [ ] Have the BL acceptance criteria been met for this change?
- [ ] Was the CircleCI build successful?
- [ ] Has the code been reviewed from a standards and best practices point of view?

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

Customer
1. Access the customer app
2. Login as a customer
3. Create a new move, create a new shipment -- VERIFY validation
4. Edit the shipment before submitting move (finish later, edit) -- VERIFY validation, and OTHER edits can still be saved without modifying `requestedPickupDate`
5. Repeat with various shipment types

Office User
1. Access the Office app
2. Login as Services Counselor
3. Create a new shipment on an existing move -- VERIFY validation
4. Edit that same move -- VERIFY validation
5. Repeat with various shipment types
6. Do the above again as a TOO

### Frontend

- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://transcom.github.io/mymove-docs/docs/adrs/Browser-Support/#minimum-browser-requirements) (Chrome, Firefox, Edge).
- [ ] There are no new console errors in the browser devtools.
- [ ] There are no new console errors in the test output.
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.
- [ ] This change meets the standards for [Section 508 compliance](https://www.ssa.gov/accessibility/andi/help/install.html).

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/getting-started/development/logging).
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/backend/guides/golang-guide#querying-the-database-safely) have been satisfied.

### Database

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#zero-downtime-migrations).
- [ ] Have been communicated to #g-database.
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#secure-migrations).

## Screenshots
<img width="561" alt="image" src="https://github.com/user-attachments/assets/7a65b5b3-ac34-415c-8c4f-525a775ae891" />

elizabeth-perkins commented on src/pages/Office/AddShipment/AddShipment.test.jsx:
```diff
@@ -209,13 +211,13 @@ describe('AddShipment component', () => {
         await user.click(screen.getByLabelText('Use pickup address'));
+        await userEvent.type(screen.getByLabelText('Requested pickup date'), tomorrow);
+        await userEvent.type(screen.getByLabelText('Requested delivery date'), '08 Nov 2020');
       });
```
I'm getting a failing client test when running this locally for this file.  States it's failing on line 228 at expect(mockNavigate).toHaveBeenCalledWith('/moves/move123/details');

danieljordan-caci replied:
Just want to chime in here - I am not seeing this test failing - might just need to re-run it
<img width="421" alt="Screenshot 2025-04-11 at 12 31 45 PM" src="https://github.com/user-attachments/assets/2704ddb5-38e2-4f11-b730-2e3c7bfefe7b" />

ryan-mchugh replied:
same here, am not able to get it failing locally, gitlab has it passing too?

elizabeth-perkins replied:
Passed with another run.  Not sure what happened the first run.

danieljordan-caci commented on src/components/Customer/MtoShipmentForm/MtoShipmentForm.test.jsx:
```diff
@@ -473,17 +476,6 @@ describe('MtoShipmentForm component', () => {
       expect(screen.getAllByLabelText(/Location Lookup/).length).toBe(4);
     });
 
-    it('goes back when the back button is clicked', async () => {
-      renderMtoShipmentForm();
-
-      const backButton = await screen.findByRole('button', { name: 'Back' });
-      await userEvent.click(backButton);
-
-      await waitFor(() => {
-        expect(mockNavigate).toHaveBeenCalledWith(-1);
-      });
-    });
```
You mean to delete this?

ryan-mchugh replied:
nope adding back, added code here, then scrapped too much

ryan-mchugh replied:
restored

danieljordan-caci reviewed (changes requested):
Any objection to using `ErrorMessage` component instead? Just matches the validation styling and looks a little nicer on mobile - I also moved it to underneath the input to keep the input from shifting down when the error shows up - just seemed less clunky.

![Screenshot 2025-04-11 at 12 37 06 PM](https://github.com/user-attachments/assets/2b9d363a-b90e-414c-9e27-db6b832d9958)

![Screenshot 2025-04-11 at 12 37 27 PM](https://github.com/user-attachments/assets/a4be17ef-a9f8-4789-86b6-ce800f8e4740)

![Screenshot 2025-04-11 at 12 37 49 PM](https://github.com/user-attachments/assets/91e6a5ae-f90c-46f8-97f4-826a3f39ca6a)

```
                            {preferredPickupDateErrorMessage && (
                              <ErrorMessage
                                id="preferredPickupDateErrorAlert"
                                data-testid="preferredPickupDateErrorAlert"
                              >
                                {preferredPickupDateErrorMessage}
                              </ErrorMessage>
                            )}
```

ryan-mchugh commented:
> Any objection to using `ErrorMessage` component instead? Just matches the validation styling and looks a little nicer on mobile - I also moved it to underneath the input to keep the input from shifting down when the error shows up - just seemed less clunky.
> 
> ![Screenshot 2025-04-11 at 12 37 06 PM](https://private-user-images.githubusercontent.com/136510600/432879742-2b9d363a-b90e-414c-9e27-db6b832d9958.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2NDQ0MDAsIm5iZiI6MTc0NDY0NDEwMCwicGF0aCI6Ii8xMzY1MTA2MDAvNDMyODc5NzQyLTJiOWQzNjNhLWI5MGUtNDE0Yy05ZTI3LWRiNmI4MzJkOTk1OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxNFQxNTIxNDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYzk5M2UwZWIzYjYyMmZkMWYyMjM3MGRmMmU0ODY4YjQzOTYyMDU1MTNkODc0MzZjMDBkZmYwYjAxMzk0YjYwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.518ox7y0e-5q-wuD3hE7IHFbpD8UOnb8ZfY-0Y5hPvg)
> 
> ![Screenshot 2025-04-11 at 12 37 27 PM](https://private-user-images.githubusercontent.com/136510600/432879710-a4be17ef-a9f8-4789-86b6-ce800f8e4740.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2NDQ0MDAsIm5iZiI6MTc0NDY0NDEwMCwicGF0aCI6Ii8xMzY1MTA2MDAvNDMyODc5NzEwLWE0YmUxN2VmLWE5ZjgtNDc4OS04NmI2LWNlODAwZjhlNDc0MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxNFQxNTIxNDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMWQwZTE1ZWYwOGI0ZDUwN2NiN2U1M2YwZTEyNmE5MTk4MjJlOGRkZDRjOTZmMGI5ZDUzNGI1ZmY5YjMyMDE1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.G11LcTQLPldQR5dQpQaTuY9h6RIfP9DqTF7Oak0vkbY)
> 
> ![Screenshot 2025-04-11 at 12 37 49 PM](https://private-user-images.githubusercontent.com/136510600/432879717-91e6a5ae-f90c-46f8-97f4-826a3f39ca6a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2NDQ0MDAsIm5iZiI6MTc0NDY0NDEwMCwicGF0aCI6Ii8xMzY1MTA2MDAvNDMyODc5NzE3LTkxZTZhNWFlLWY5MGMtNDZmOC05N2Y0LTgyNmEzZjM5Y2E2YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxNFQxNTIxNDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ZjZkMWRjOTRlZDdjNzg3YzMxMDE2MGQyODA0OTA2ODIzNDc1MTJkN2Q3Njc5NWU4NzZlMTY1NGQ1MWRhM2UzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.-yNaxSCsZypmhva7xM12QyC5FvPTCx8m-Ds1VV4VQss)
> 
> ```
>                             {preferredPickupDateErrorMessage && (
>                               <ErrorMessage
>                                 id="preferredPickupDateErrorAlert"
>                                 data-testid="preferredPickupDateErrorAlert"
>                               >
>                                 {preferredPickupDateErrorMessage}
>                               </ErrorMessage>
>                             )}
> ```

Done! Used the <ErrorMessage> component and updated tests to support. Had to wrap in a "span" to get the TestId passed, let me know if you know a better solution for that to pass through.

danieljordan-caci reviewed (commented):
![Screenshot 2025-04-14 at 1 08 41 PM](https://github.com/user-attachments/assets/60604795-75a3-437a-8417-3c550cea9f00)
![Screenshot 2025-04-14 at 1 09 48 PM](https://github.com/user-attachments/assets/df190478-3090-40ad-a544-af51ebd01725)
![Screenshot 2025-04-14 at 1 09 54 PM](https://github.com/user-attachments/assets/8411986d-826b-421d-98fc-165fe2e09686)


Looks good on desktop and mobile! 

One question... If we are not allowing a date to be selected, should we even be showing the weekend/holiday blurb?

ryan-mchugh commented:
@danieljordan-caci Hides the holiday alert when the requestedPickupDate validation error in the past is present. Added that workflow to existing test

danieljordan-caci reviewed (approved):
Looks great! All tests are passing, functionality is rock solid. Appreciate the changes!

https://github.com/user-attachments/assets/b40b31b3-541e-47d1-823c-4708a4b84863

samaysofo commented:
works fine for me,  just a thought showing both "Required" and "Requested pickup date must be in the future"  error messages at same time seems not accurate to me. When the field do have a date i would think the "Required" error is irrelevant at that point and the most relevant message is the "Requested pickup date must be in the future".
What i would expect to see as a user is if no dates entered "Required" error and if dates entered but doesn't meet the requirements then "Requested pickup date must be in the future".

ryan-mchugh commented:
> works fine for me, just a thought showing both "Required" and "Requested pickup date must be in the future" error messages at same time seems not accurate to me. When the field do have a date i would think the "Required" error is irrelevant at that point and the most relevant message is the "Requested pickup date must be in the future". What i would expect to see as a user is if no dates entered "Required" error and if dates entered but doesn't meet the requirements then "Requested pickup date must be in the future".

Let me see what I can do.

danieljordan-caci commented:
> > works fine for me, just a thought showing both "Required" and "Requested pickup date must be in the future" error messages at same time seems not accurate to me. When the field do have a date i would think the "Required" error is irrelevant at that point and the most relevant message is the "Requested pickup date must be in the future". What i would expect to see as a user is if no dates entered "Required" error and if dates entered but doesn't meet the requirements then "Requested pickup date must be in the future".
> 
> Let me see what I can do.

You'll have to address the validation schema. I'm pretty sure you can achieve this with most of your logic being in the schema.

ryan-mchugh commented:
Still have some more changes, not sure where my head was at, but we do NOT want the validation running if it was not touched, to allow for edits.

danieljordan-caci commented:
> Still have some more changes, not sure where my head was at, but we do NOT want the validation running if it was not touched, to allow for edits.

@ryan-mchugh is this ready for review again or are you actively making changes?

ryan-mchugh commented:
> > Still have some more changes, not sure where my head was at, but we do NOT want the validation running if it was not touched, to allow for edits.
> 
> @ryan-mchugh is this ready for review again or are you actively making changes?

still making changes

danieljordan-caci commented:
> > > Still have some more changes, not sure where my head was at, but we do NOT want the validation running if it was not touched, to allow for edits.
> > 
> > 
> > @ryan-mchugh is this ready for review again or are you actively making changes?
> 
> still making changes

Okay. Can you change this to draft status then? Thanks!

ryan-mchugh commented:
@danieljordan-caci It's updated and ready for review again!

danieljordan-caci reviewed (approved):
Looks good!

samaysofo commented:
> > Still have some more changes, not sure where my head was at, but we do NOT want the validation running if it was not touched, to allow for edits.
> 
> @ryan-mchugh is this ready for review again or are you actively making changes?

LGTM, thanks for the updates 👏
